### PR TITLE
Optimize Trie Snapshot and Benchmark against deep copy.

### LINKED ISSUE #1343: dot/state: store trie nodes individually in database
## Task summary

- currently, once a block is imported, the entire state trie is encoded and stored in the database at the root hash as the key
- this is very inefficient since most of the keys don't change between blocks
- instead, the trie should be stored in the database by trie node
- key = hash of trie node; value = encoded trie node
- this way, the trie can be recursively reconstructed by looking up its root hash, decoding the trie node from there, and then looking up its children nodes by hash, and so on until the trie in reconstructed
- to update a trie value, only the node that is updated and its parents up to the root need to be updated

Note: trie node keys in database should be prefixed by some prefix (`trie` perhaps) to allow for the keys to have improved locality within the database

need to implement the following:
```
type Trie interface {
     // Store stores each trie node in the database, where the key is the hash of the encoded node and the value is the encoded node. Generally, this will only be used for the genesis trie.
     Store(db chaindb.Database) error

     // Load reconstructs the trie from the database from the given root hash. Used when restarting the node to load the current state trie.
     Load(db chaindb.Database, root common.Hash) error

     // PutInDB puts a value into the trie and writes the updates nodes the database. Since it needs to write all the nodes from the changed node up to the root, it writes these in a batch operation.
     PutInDB(db chaindb.Database, key, value []byte) error 

     // DeleteFromDB deletes a value from the trie and writes the updated nodes the database. Since it needs to write all the nodes from the changed node up to the root, it writes these in a batch operation.
     DeleteFromDB(db chaindb.Database, key []byte) error

     // ClearPrefixFromDB deletes all keys with the given prefix from the trie and writes the updated nodes the database. Since it needs to write all the nodes from the changed node up to the root, it writes these in a batch operation.
     ClearPrefixFromDB(db chaindb.Database, key []byte) error

     // GetFromDB retrieves a value from the trie using the database. It recursively descends into the trie using the database starting from the root node until it reaches the node with the given key. It then reads the value from the database.
     GetFromDB(db chaindb.Database, key []byte) ([]byte, error)
}
```

``` 
// StorageState is used to read and write storage tries from the database, as well as read storage values given a root hash.
// It should also store the current storage root at a database key. 
type StorageState struct {
    db chaindb.Database 
    currentTrie Trie // the current storage trie at the chain head
}

type StorageState interface {
    // Loads the current trie from the database by first reading the current storage root from the db, then reconstrucitng the trie. 
    LoadCurrentTrie() error 

     // Get returns a storage value given the trie root hash. It recursively descends into the trie using the database starting from the root node until it reaches the node with the given key. It then reads the value from the database.
    Get(root common.Hash, key []byte) ([]byte, error)

    // NewRuntimeStorage returns a new RuntimeStorage from the given storage root.  If the root is the current root, it uses the trie cached in the StorageState after copying it; otherwise, it (optionally) loads it from the database. (note: may not be needed to load the entire trie, could just pass in the database.)
    NewRuntimeStorage(root common.Hash) RuntimeStorage
}
```

```
type RuntimeStorage type {
      db chaindb.Database
      trie Trie
}

// RuntimeStorage is the interface used by the runtime to read and write the storage.
type RuntimeStorage interface {
      // Same as current runtime.Storage interface
}
```

see https://github.com/ChainSafe/gossamer/pull/1341 for preliminary benchmarking

### PR DESCRIPTION
**Pull request by arijitAD**
## Changes

- Implement versioned trie and benchmark.

### Benchmark Results (Snapshot vs Deepcopy)
```
Number of iteration = 1000
```
#### Snapshot
```
=== RUN   TestDeepCopy/Snapshot
2021/02/09 20:09:25 
Alloc = 38 MB
TotalAlloc = 53 MB 
Sys = 71 MB 
NumGC = 4 
2021/02/09 20:09:25 Snapshot to trie took 31.693922ms
```
#### Deepcopy
```
=== RUN   TestDeepCopy/DeepCopy
2021/02/09 20:09:57
Alloc = 4869 MB
TotalAlloc = 4915 MB
Sys = 5093 MB
NumGC = 12
2021/02/09 20:09:57 DeepCopy to trie took 5.219532653s
```

## Tests

```
TestSnapshot
TestDeepCopyVsSnapshot
```

## Checklist

- [x] I have read [CODE_OF_CONDUCT](https://github.com/ChainSafe/gossamer/blob/development/.github/CODE_OF_CONDUCT.md) and [CONTRIBUTING](https://github.com/ChainSafe/gossamer/blob/development/.github/CONTRIBUTING.md) 
- [x] I have provided as much information as possible and necessary
- [x] I have reviewed my own pull request before requesting a review
- [ ] All integration tests and required coverage checks are passing

## Issues

- closes #1343

### CODE REVIEW

**Review by RyRy79261 — CHANGES_REQUESTED** [review]
`TestDeepCopyVsSnapshot` times out: 
image

File: lib/trie/trie.go [hunk 1]
```diff
@@ -65,6 +102,46 @@ func (t *Trie) DeepCopy() (*Trie, error) {
...
+			return bytes.Equal(tLeaf.value, trieLeaf.value)
+		}
+		return false
+	}
+	return false
+}
+
+// Equal returns true if both tries are same.
+func (t *Trie) Equal(trie *Trie) bool {
+	return isSimilar(t.root, trie.root)
```
Comment (noot) [inline lib/trie/trie.go:142]: why is this needed? can we not just compare the root hashes?

File: lib/trie/trie.go [hunk 2]
```diff
@@ -32,31 +32,68 @@ var EmptyHash, _ = NewEmptyTrie().Hash()
 // The zero value is an empty trie with no database.
 // Use NewTrie to create a trie that sits on top of a database.
 type Trie struct {
-	root     node
-	children map[common.Hash]*Trie
+	generation uint64
+	root       node
+	children   map[common.Hash]*Trie // Used to store the child trie.
```
Comment (noot) [inline lib/trie/trie.go:37]: child tries*

**Review by edwardmack — APPROVED** [review]
Nice work, On my machine Snapshot is about 150 times faster than DeepCopy, but they both used the same memory:
```
=== RUN   TestDeepCopyVsSnapshot/DeepCopy
2021/02/09 17:24:27 
Alloc = 4868 MB 
TotalAlloc = 4914 MB 
Sys = 5093 MB 
NumGC = 12 

2021/02/09 17:24:27 DeepCopy to trie took 1.563547599s
    --- PASS: TestDeepCopyVsSnapshot/DeepCopy (1.56s)
=== RUN   TestDeepCopyVsSnapshot/Snapshot
2021/02/09 17:24:27 
Alloc = 4868 MB 
TotalAlloc = 4915 MB 
Sys = 5093 MB 
NumGC = 12 

2021/02/09 17:24:27 DeepCopy to trie took 9.643327ms
```

File: lib/trie/trie.go [hunk 3]
```diff
@@ -225,30 +302,38 @@ func (t *Trie) Put(key, value []byte) error {
...
+		if p != nil && n != nil && n.isDirty() {
+			p.setDirty(true)
+		}
+		return n
 	case nil:
+		// We are creating new node so it will always have the latest generation.
+		// TODO: Sanity check. Can be removed later.
+		if t.generation != value.getGeneration() {
+			log.Error("node with invalid generation inserted")
+		}
```
Comment (noot) [inline lib/trie/trie.go:325]: pls remove if no longer needed

File: lib/trie/trie.go [hunk 4]
```diff
@@ -225,30 +302,38 @@ func (t *Trie) Put(key, value []byte) error {
...
 		switch v := value.(type) {
 		case *branch:
 			v.key = key
-			n = v
+			return v
```
Comment (noot) [inline lib/trie/trie.go:329]: nice :D

File: dot/utils_test.go [hunk 5]
```diff
@@ -86,3 +90,105 @@ func TestNewRuntimeFromGenesis(t *testing.T) {
 	_, err := wasmer.NewRuntimeFromGenesis(gen, &wasmer.Config{})
 	require.NoError(t, err)
 }
+
+func TestDeepCopyVsSnapshot(t *testing.T) {
```
Comment (noot) [inline dot/utils_test.go:94]: is it possible to change this into a test benchmark? eg. something like this https://golang.org/pkg/testing/#hdr-Benchmarks

**Review by noot — COMMENTED** [review]
great work! nice cleanups :D i think you can probably remove the `trie.Equal` function - comparing root hashes should tell you if the tries are the same, since that's the point of being able to hash the trie.

File: lib/trie/trie.go (hunk 1 above)
Comment (arijitAD) [reply lib/trie/trie.go:142]: Silly me 😅 It skipped my mind that the trie is also a Merkle tree.

**Comment by arijitAD** [comment]
> Nice work, On my machine Snapshot is about 150 times faster than DeepCopy, but they both used the same memory:
> 
> ```
> === RUN   TestDeepCopyVsSnapshot/DeepCopy
> 2021/02/09 17:24:27 
> Alloc = 4868 MB 
> TotalAlloc = 4914 MB 
> Sys = 5093 MB 
> NumGC = 12 
> 
> 2021/02/09 17:24:27 DeepCopy to trie took 1.563547599s
>     --- PASS: TestDeepCopyVsSnapshot/DeepCopy (1.56s)
> === RUN   TestDeepCopyVsSnapshot/Snapshot
> 2021/02/09 17:24:27 
> Alloc = 4868 MB 
> TotalAlloc = 4915 MB 
> Sys = 5093 MB 
> NumGC = 12 
> 
> 2021/02/09 17:24:27 DeepCopy to trie took 9.643327ms
> ```

I forgot to trigger garbage collection after each test. It should return the correct stats now.

File: dot/utils_test.go (hunk 5 above)
Comment (arijitAD) [reply dot/utils_test.go:94]: We can but for loading genesis file the code is tightly coupled to use `testing.T` struct. We can modify the function to take `require.TestingT` interface as input but it will be major refactor. I can do this in a separate PR.

**Comment by arijitAD** [comment]
> `TestDeepCopyVsSnapshot` times out:
> image

Screenshot 2021-02-10 at 5 03 17 PM
It works for me. Can you remove the timeout and use the above command?

File: lib/trie/trie.go [hunk 6]
```diff
@@ -230,30 +260,34 @@ func (t *Trie) Put(key, value []byte) error {
...
 	k := keyToNibbles(key)
 
-	n, err := t.insert(t.root, k, &leaf{key: nil, value: value, dirty: true, valueDirty: true})
-	if err != nil {
-		return err
-	}
-
-	t.root = n
+	t.root = t.insert(t.root, k, &leaf{key: nil, value: value, dirty: true, valueDirty: true, generation: t.generation})
 	return nil
```
Comment (noot) [inline lib/trie/trie.go:264]: since the error returned here is always nil, you can remove the return param

**Review by noot — APPROVED** [review]
awesome work! only 1 small comment, everything else looks great.
